### ISSUE TITLE
glasnt tries out django-simple-deploy on fly.io

### ISSUE THREAD

**Issue by glasnt** [ehmatthes/django-simple-deploy#156]
Since I called the last issue `heroku`, a separate one for flyio. 

It worked well, but again, I had to work out my own staticfiles and createsuperuser pathway.

## createsuperuser

From a hint in https://usher.dev/posts/wagtail-on-flyio/part-5/ I found `flyctl ssh console`. Personally first time I ran it I got `Error error connecting to SSH server`, but immediately running it again I got a console. I could then use the interactive `createsuperuser`

## static

There was a note in the Dockerfile that ` # This is in the default Fly.io Dockerfile, but doesn't work with the current workflow used by simple_deploy.` (only from 10 days ago #116)

For my setup, I ended up adding whitenoise similar to what I did in #152, but with a twist: adding the staticfiles configs to be used on both a `ON_FLYIO` and `ON_FLYIO_SETUP` envvar, then having the Dockerfile run with this setup envvar: 

Dockerfile: 
```diff
-#RUN python manage.py collectstatic --noinput
+RUN ON_FLYIO_SETUP=true python manage.py collectstatic --noinput
```

setup.py append
```
if os.environ.get("ON_FLYIO_SETUP") or os.environ.get("ON_FLYIO"):
     # from https://whitenoise.evans.io/en/stable/#quickstart-for-django-apps
     STATIC_ROOT = os.path.join(BASE_DIR, 'staticfiles')
     STATIC_URL = '/static/'
     STATICFILES_DIRS = (os.path.join(BASE_DIR, 'static'),)
     i = MIDDLEWARE.index("django.middleware.security.SecurityMiddleware")
     MIDDLEWARE.insert(i + 1,"whitenoise.middleware.WhiteNoiseMiddleware")
```

This way, you can the collectstatic command in the Dockerfile even without the database connectivity (thanks to a hint from @kenwhitesell over in https://forum.djangoproject.com/t/django-collectstatic-doesnt-work-without-a-db/4098/12) but you have these settings on both fly setup (since dockerfile in fly's build process doesn't have access to secrets) and fly itself (which has the 'secret')

Heroku runs this command itself, but I guess manually running this in Dockerfile is okay

You could also adjust the `release_command` to execute both `migrate` and `collectstatic`, but you'd need to ensure the static settings were available outside the `ON_FLYIO` conditional as that is also not available in that execution environment.

Fly also doesn't have Django docs, but it has Python docs https://fly.io/docs/languages-and-frameworks/python/ which mention Procfiles, but  per #116 there's an 'official Dockerfile' (though I don't see this mentioned on https://fly.io/docs/languages-and-frameworks/dockerfile/ so unsure where this file lives normally, but this should be documented so it can be kept up to date. (Or, dropped for Procfile if the things the Dockerfile adds can be moved elsewhere, depending on what bits the Dockerfile brings apart from the extra collectstatic step)

**Comment by ehmatthes** [comment]
All right, this is really helpful; #157 tracks the changes needed to address this.

Fly staff have acknowledged that their initial Django example was not well informed, and their Python docs are not thorough. They should have some new Django documentation coming out shortly, and that documentation incorporates feedback from people in the Django community. I will be curious to see exactly what that a new documentation looks like, and how their example project and Dockerfiles have been modified.

I wouldn't ever imagine that simple_deploy would create a superuser for people. There are two places we can address that if it's helpful. We could put that in the success message that's generated at the end of a configuration run and an automate_all run. That's probably a bit much to include there. The heroku configuration generates a stub friendly log in the simple_deploy_logs directory. That friendly log can contain much more information than the success message. It should not try to recreate too much of the platform's documentation; it's meant to be a bridge to the platform's documentation for people unfamiliar with the docs, and maybe a cheat sheet for people who have some familiarity with the platform.

When the friendly log has been built out, the final success message will provide a clear suggestion to go look at the friendly log. In the automate_all mode, we can also consider opening both the deployed project in a new browser tab (this already happens), and the friendly log as well.

**Comment by glasnt** [comment]
>  I will be curious to see exactly what that a new documentation looks like, and how their example project and Dockerfiles have been modified.

I would also be interested. Can you let me know when you see an update? Thanks :)

> I wouldn't ever imagine that simple_deploy would create a superuser for people

No, but I think it should include after output information about how to do that. Even if it's just a link to the `ssh` equiv for the platform, and maybe a `[ssh invoke] python manage.py createsuperuser` suggestion. 

In my samples I often have a data migration where it uses a `ADMIN_PASSWORD` envvar to seed the superuser, just to skip that step.

**Comment by ehmatthes** [comment]
> I would also be interested. Can you let me know when you see an update? Thanks :)

Certainly.

**Comment by ehmatthes** [comment]
> Fly also doesn't have Django docs, but it has Python docs https://fly.io/docs/languages-and-frameworks/python/ which mention Procfiles, but per https://github.com/ehmatthes/django-simple-deploy/pull/116 there's an 'official Dockerfile' (though I don't see this mentioned on https://fly.io/docs/languages-and-frameworks/dockerfile/ so unsure where this file lives normally, but this should be documented so it can be kept up to date. (Or, dropped for Procfile if the things the Dockerfile adds can be moved elsewhere, depending on what bits the Dockerfile brings apart from the extra collectstatic step)

Here is a [PR](https://github.com/superfly/flyctl/pull/1366) where the flyio dockerfile was updated with feedback from some Django folks. And here is the [dockerfile](https://github.com/superfly/flyctl/blob/master/scanner/templates/django/Dockerfile) that's generated when you run `fly launch` for a Django project. I don't know that this file is posted anywhere other than here in the CLI source.

I'm planning to write up a detailed description of the deployment process for each platform, with references to the platform docs. This kind of link will go in that documentation.

**Comment by ehmatthes** [comment]
I believe this is resolved in 0.5.9. For more details, see #157. Here's the current [Dockerfile](https://github.com/ehmatthes/django-simple-deploy/blob/main/simple_deploy/templates/dockerfile_flyio) and [settings block](https://github.com/ehmatthes/django-simple-deploy/blob/main/simple_deploy/templates/flyio_settings.py).
